core/src/main/java/org/apache/calcite/rel/core/Join.java

```diff
@@ -230,6 +237,17 @@ public boolean isSemiJoinDone() {
     return false;
   }
 
+  /**
+   * Returns whether this LogicalJoin is a semi-join but does
+   * not comes from decorrelate.
+   *
+   * @return true if this is semi but without correlate variables.
+   */
+  public boolean isNonCorrelateSemiJoin() {
+    return (this.variablesSet == null || this.variablesSet.size() == 0)
+        && joinType == JoinRelType.SEMI;
+  }
```

Comment:
The addition of this method confuses me.

Reply:
In the beginning, i add this method to distinguish the `SemiJoin` transformed from `Correlate` and `SemiJoin` from `SemiJoinRule` or `RelBuilder`. Cause i thought that the join comes from a `Correlate` can make a JoinRelType.SEMI join type.

But in current code, Calcite will never transform from Correlate to join with `JoinRelType.SEMI`(only left and inner join type), so this method actually can simplify to just decide if the join type is `JoinRelType.SEMI`.

But i still keep the variables check and i think correlate should have the ability to get a semi join type join.

Reply:
A join should not be correlated. Only Correlated is allowed to be correlated, in theory.

Reply:
I can’t imagine how to achieve a nested-loop join without a correlating variable. The left side sets it for each row, and the right side restarts and applies some condition that uses the variable. So we’re just arguing over nomenclature.

Reply:
Another reason i see is that Calcite supports converting from a Correlate to Join (both for logical nodes and physical), i can't see another way how to keep these variables into a transformed join.

If Correlate is just a correlate and have it's self implementation, the concept is more clear, and i supports @hsyuan 's idea but we don't now.

Reply:
Currently we have two concepts `Join` and `Correlate`. If we say that the `Join` can have correlated variables then I don't see why we need to keep `Correlate`.  
The discussion so far (and the current PR) proposes to keep `Correlate` so that's why I think that `Join` should not deal with correlated variables.

> Calcite supports converting from a Correlate to Join ...

@danny0405 can you explain a bit more where this is happening? I know that the opposite is done with JoinToCorrelateRule but I don't remember seeing CorrelateToJoinRule (or something similar). We can certainly add such a rule but I argue that in this case we should be able to get rid of correlated variables.

Reply:
@danny0405 The purpose of transforming a correlate to a join is getting rid of Variables.
@zabetak It is in RelDecorrelator, which is not rule based, sadly. Some SubqueryRemoveRule will transform subquery into Correlate, some SubqueryRemoveRule will directly transform subquery into a Join. :(

Reply:
NestedLoop can represent both Join and Correlate. Say for query
```
select * from R, S where r1 > s1;
```

We can have plan:
```
NLJ (r1 > s1)
 +- R
 +- S

for (r in R) {
  for (s in S) {
    if (condition(r, s) is true)
         output r,s;
  }
}
```

or 
```
NLJ (true)
  +- R
  +- Filter (r1 > s1)
        +- S

for (r in R) {
  while (s = scanNext(innerRel, r))
     output r,s;
}
```

The difference is can we start fetching inner tuple without knowing the tuple from outer. It is impossible for the 2nd, which is a Correlate implementation. It is especially true if the inner is a index scan.

It is true that we can always use the second one to implement it, depending on the implementation, that's why we have JoinToCorrelate rule.

But logically I think we'd better not mix them.

Reply:
Thanks for the clarification @hsyuan. What I was wondering actually is if we really use `variablesSet` in the `Join` operator? As far as I can see at least in our test suite it is never used. If there are variables to set we currently use only `Correlate` . So I was thinking is that if we keep both `Correlate` and `Join`  there is no reason to have `variablesSet` in the `Join` operator.

Reply:
variablesSet is not used in Join, and shouldn't be used. I agree with you, no need to have variablesSet in Join operator.

Reply:
I agree to remove variables also, as far as i can see, the variables are used when building a Correlate[1], If we see Correlate as (Join + variableSet), remove variableSet from Join may be the more correct way to go.

Another thing i see is that the Correlate do not have join condition, which eliminate it's ability.

[1]https://github.com/apache/calcite/blob/a3f81bb7b088fd8c1d0c1df3b0f2b0cf122633de/core/src/main/java/org/apache/calcite/tools/RelBuilder.java#L1729

Reply:
Correlate's condition is always TRUE, which is expected. What do you mean by `eliminate it's ability.`?

Reply:
I mean the Correlate actually could have a filter condition, which can be pushed into it.

Reply:
@hsyuan Removing variables set from Correlate is another huge change, if we really decide to do it, i'm suggesting to put it in another patch. And i'm glad to take it.

Reply:
We don't need to decorrelate anything here in this patch, RelDecorrelator already does that work.

Reply:
Wow, my mistake, i was talking about removing variablesSet from Join.

Reply:
We can do it in a different patch.

Reply:
@hsyuan  I have rename the method name to isSemiJoin and remove check for variablesSet, i didn't add assert cause it is not needed, This is more clear after we remove variablesSet from Join.

Reply:
The reason for variablesSet is correlated scalar sub-query in the ON clause. Similarly Filter needs variablesSet if there is a correlated scalar sub-query in its condition, and Project needs variablesSet for correlated scalar sub-query in one of the project expressions.

This stuff isn't very well tested, but it is needed because sub-query can occur anywhere in a query.